# 浏览器环境下的微任务
MutationObserver接口 和 Promise 在浏览器换下都会开启微任务,Promise 在开发中使用的较多,下面我们来了解下;
[MutationObserver MDN](https://developer.mozilla.org/zh-CN/docs/Web/API/MutationObserver);[Promise MDN](https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Promise)

## Promise
先来看看[MDN promise的使用](https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Guide/Using_promises);
#### 引入 promise 的背景
js 的单线程运行机制决定了 js在处理耗时任务时采用`异步回调`的方式, 那么在处理异步任务时序的时候就会产生后续的任务`依赖`上一个任务的结果数据(结果还存在`不确定性`的问题), 我们通常的解决方式就回调嵌套回调, 这样又会导致我们的编码方式异常难以理解和维护...
为此, 引入了promise 来消除嵌套调用和以及任务的错误处理这些问题.
#### promise 的使用
```js
function XFetch(request) {
  function executor(resolve, reject) {
      let xhr = new XMLHttpRequest()
      xhr.open('GET', request.url, true)
      xhr.ontimeout = function (e) { reject(e) }
      xhr.onerror = function (e) { reject(e) }
      xhr.onreadystatechange = function () {
          if (this.readyState === 4) {
              if (this.status === 200) {
                  // resolve调用时, 会触发下面 promise.then 的回调 
                  resolve(this.responseText, this)
              } else {
                  let error = {
                      code: this.status,
                      response: this.response
                  }
                  // resolve调用时, 会触发下面 promise.catch 的回调 
                  reject(error, this)
              }
          }
      }
      xhr.send()
  }
  return new Promise(executor)
}

// 写法更线性,优美
// function request: to make a request config object
var x1 = XFetch(request('first url'))

var x2 = x1.then(value1 => {
    console.log(value1)
    return XFetch(request('second url'))
})

var x3 = x2.then(value2 => {
    console.log(value2)
    return XFetch(request('third url'))
})

x3.catch(error => {
    console.log(error)
})
```
#### promise 的消除回调嵌套的机理
- 为什么要使用函数嵌套
```
一个主要原因是在发起任务请求时会带上回调函数,这样当任务处理结束之后,
下个任务就只能在回调函数中来处理了, 那我们试试将回调函数存起来怎么样?
```
- Promise 实现了回调函数的延时绑定
```js
// 业务函数
function executor(resolve, reject){
    setTimeout(() => {
        resolve(100)
    }, 1000)
}
// 回调函数
function onResolve(value){
    console.log(value)
}

// 创建Promise对象p1,并执行executor函数
let p1 = new Promise(executor)
// p1.then设置回调函数onResolve
p1.then(onResolve)
```
- Promise 将回调函数 onResolve 的返回值穿透到最外层  
根据 onResolve 函数的传入值来决定创建什么类型的 Promise 任务,
将创建好的 Promise 对象需要返回到最外层,这样就可以摆脱嵌套循环
```js
// 业务函数
function executor(resolve, reject){
    setTimeout(() => {
        resolve(100)
    }, 1000)
}
let p11

// 回调函数
function onResolve(value){
    console.log(value)
    // 再创建一个 promise 实例
    let p = new Promise((resolve, reject) => {
        resolve(value + 100)
    })
    p11 = p
    console.log('实例inner', p)
    // 返回这个实例
    return p
}
// 创建Promise对象p1,并执行executor函数
let p1 = new Promise(executor)
// p1设置回调函数onResolve
let p2 = p1.then(onResolve)
p2.then(value2 => {
    console.log(value2)
    console.log('实例outer', p2)
})
```
- Promise 对象的错误具有“冒泡”性质,会一直向后传递,直到最后被catch
```js
function executor(resolve, reject) {
    let rand = Math.random();
    console.log(1)
    console.log(rand)
    if (rand > 0.5)
        resolve()
    else
        reject()
}
var p0 = new Promise(executor);

var p1 = p0.then((value) => {
    console.log("succeed-1")
    return new Promise(executor)
})

var p3 = p1.then((value) => {
    console.log("succeed-2")
    return new Promise(executor)
})

var p4 = p3.then((value) => {
    console.log("succeed-3")
    return new Promise(executor)
})

p4.catch((error) => {
    console.log("error")
})
console.log(2)
// p0～p4。无论哪个对象里面抛出异常,都可以通过最后一个对象 p4.catch 来捕获异常,
// 通过这种方式可以将所有 Promise 对象的错误合并到一个函数来处理;
// 这样就解决了每个任务都需要单独处理异常的问题

// 怎么实现的呢?
// 判断then中没有第二个参数处理异常,
// 有则包装成promise.reject形式返回
// 没有则继续返回promise.reject的promise直到被处理
```
- 极简版 Promise  
[promise源码参考](https://github.com/then/promise/tree/master/src)

```js
/*
 * 1.回调函数延迟执行
*/
function MyPromise(executor) {
    var _resolve = null
    var _reject = null
    // 模拟实现then
    this.then = function (resolve, reject) {
        _resolve = resolve
    };
    function resolve(data) {
        // 延迟执行
        setTimeout(()=>{
            _resolve(data)
        }, 0)
    }
    executor(resolve, null);
}

function func(resolve, reject) {
    resolve(1)
}
let p = new MyPromise(func)
p.then(data => {
    console.log(data)
})
```

## 小结
Promise的产生主要是为了解决:
- 回调地狱问题
- 从异步编程的代码不连续问题
- 封装异步代码,让处理流程变的线性
- 在复杂情况下引出了嵌套调用、任务不确定性两个问题
- Promise通过实现函数的`延时绑定`和`回调函数的返回值穿透到最外面`来解决上面问题